File: Source/Scene.cpp
#include "Eigen/Core"
#include "Model.hpp"
#include "global.hpp"
#include "light.hpp"
#include <Scene.hpp>
#include <algorithm>
#include <cmath>
#include <functional>
#define STB_IMAGE_IMPLEMENTATION
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include <stb_image.h>
#include <stb_image_write.h>

Scene::Scene(int width, int height)
    : eye_pos{0.0f, 0.0f, 0.0f}, view_dir{0.0f, 0.0f, -1.0f}, fov(45.0f),
      aspect_ratio(1.0f), zNear(-0.1f), zFar(-1000.0f), width(width),
      height(height) {}

void Scene::start_render() {
  frame_buffer.resize(width * height, {0.7f, 0.7f, 0.7f});
  pos_buffer.resize(width * height, {0.0f, 0.0f, 0.0f});
  normal_buffer.resize(width * height, {0.0f, 0.0f, 0.0f});
  diffuse_buffer.resize(width * height, {0.0f, 0.0f, 0.0f});
  specular_buffer.resize(width * height, {0.0f, 0.0f, 0.0f});
  glow_buffer.resize(width * height, {0.0f, 0.0f, 0.0f});
  z_buffer.resize(width * height, INFINITY);
  std::fill(frame_buffer.begin(), frame_buffer.end(),
            Eigen::Vector3f{0.7f, 0.7f, 0.7f});
  std::fill(z_buffer.begin(), z_buffer.end(), INFINITY);
  Eigen::Matrix<float, 4, 4> model = Eigen::Matrix<float, 4, 4>::Identity(),
                             view = get_view_matrix(eye_pos, view_dir),
                             projection = get_projection_matrix(
                                 fov, aspect_ratio, zNear, zFar);
  Eigen::Matrix<float, 4, 4> mvp = projection * view * model;
  Eigen::Matrix<float, 4, 4> mv = view * model;
  for (auto &&light : lights) {
    light->look_at(*this);
  }
  for (auto &&obj : objects) {
    obj->clip(mvp, mv);
    obj->to_NDC(width, height);
  }
#pragma omp parallel for collapse(2)
  for (int j = 0; j < height; j += tile_size) {
    for (int i = 0; i < width; i += tile_size) {
      for (auto &&obj : objects) {
        obj->rasterization_block(*this, j, i, std::min(tile_size, height - j),
                                 std::min(tile_size, width - i));
      }
    }
  }
  int box_radius = roundf(4.0f * std::max(width, height) / 1024.0f);
  for (auto &&light : lights) {
    light->generate_penumbra_mask(*this);
    light->box_blur_penumbra_mask(box_radius);
  }
#pragma omp parallel for collapse(2)
  for (int j = 0; j < height; j += tile_size) {
    for (int i = 0; i < width; i += tile_size) {
      for (auto &&obj : objects) {
        shader(*this, j, i, std::min(tile_size, height - j),
               std::min(tile_size, width - i));
      }
    }
  }
}

void Scene::add_model(const std::shared_ptr<Model> &model) {
  objects.push_back(model);
}

void Scene::add_light(const std::shared_ptr<light> &light) {
  lights.push_back(light);
}

void Scene::delete_model(const std::shared_ptr<Model> &model) {
  objects.erase(std::remove(objects.begin(), objects.end(), model),
                objects.end());
}
void Scene::delete_light(const std::shared_ptr<light> &light) {
  lights.erase(std::remove(lights.begin(), lights.end(), light), lights.end());
}

void Scene::save_to_file(std::string filename) {
  std::vector<unsigned char> data(width * height * 3);
  for (int y = 0; y != height; ++y) {
    for (int x = 0; x != width; ++x) {
      for (int i = 0; i != 3; ++i) {
        data[3 * (y * width + x) + i] =
            roundf(std::clamp(frame_buffer[width * (height - y - 1) + x][i],
                              0.0f, 1.0f) *
                   255);
      }
    }
  }
  stbi_write_png(filename.c_str(), width, height, 3, data.data(), width * 3);
}

int Scene::get_index(int x, int y) const { return width * y + x; }

void Scene::set_eye_pos(const Eigen::Vector3f eye_pos) {
  this->eye_pos = eye_pos;
}

void Scene::set_view_dir(const Eigen::Vector3f view_dir) {
  this->view_dir = view_dir.normalized();
}
void Scene::set_fov(float fov) { this->fov = fov; }

void Scene::set_aspect_ratio(float aspect_ratio) {
  this->aspect_ratio = aspect_ratio;
}

void Scene::set_zNear(float zNear) { this->zNear = zNear; }

void Scene::set_zFar(float zFar) { this->zFar = zFar; }

void Scene::set_width(int width) { this->width = width; }

void Scene::set_height(int height) { this->height = height; }

void Scene::set_shader(
    const std::function<void(Scene &Scene, int start_row, int start_col,
                             int block_row, int block_col)> &shader) {
  this->shader = shader;
}

Eigen::Vector3f Scene::get_eye_pos() const { return eye_pos; }

Eigen::Vector3f Scene::get_view_dir() const { return view_dir; }

float Scene::get_fov() const { return fov; }

float Scene::get_aspect_ratio() const { return aspect_ratio; }

float Scene::get_zNear() const { return zNear; }

float Scene::get_zFar() const { return zFar; }

int Scene::get_width() const { return width; }

int Scene::get_height() const { return height; }
std::function<void(Scene &Scene, int start_row, int start_col, int block_row,
                   int block_col)>
Scene::get_shader() const {
  return shader;
}

File: render-lib/src/rasterization/scene.rs
use nalgebra::clamp;
use std::f32::INFINITY;
use std::sync::Arc;
use std::sync::Mutex;
use std::thread;

use nalgebra::max;

use crate::rasterization::light::*;
use crate::rasterization::model::*;
use crate::rasterization::unsafe_pack::*;
use crate::util::math::*;
const SCENE_MAXIMUM_THREAD_NUM: i32 = 8;

pub struct Scene {
    pub(in crate::rasterization) eye_pos: Vector3f,
    pub(in crate::rasterization) view_dir: Vector3f,
    pub(in crate::rasterization) fov: f32,
    pub(in crate::rasterization) aspect_ratio: f32,
    pub(in crate::rasterization) z_near: f32,
    pub(in crate::rasterization) z_far: f32,
    pub(in crate::rasterization) width: i32,
    pub(in crate::rasterization) height: i32,
    pub(in crate::rasterization) frame_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) pos_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) normal_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) diffuse_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) specular_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) glow_buffer: Vec<Vector3f>,
    pub(in crate::rasterization) z_buffer: Vec<f32>,
    pub(in crate::rasterization) models: Vec<Arc<Mutex<Model>>>,
    pub(in crate::rasterization) lights: Vec<Arc<Mutex<dyn LightTrait>>>,
    pub(in crate::rasterization) shader:
        unsafe fn(*mut Scene, *const Vec<*const dyn LightTrait>, i32, i32, i32, i32),
}
impl Default for Scene {
    fn default() -> Self {
        Scene {
            eye_pos: Vector3f::new(0.0, 0.0, 0.0),
            view_dir: Vector3f::new(0.0, 0.0, -1.0),
            fov: 45.0,
            aspect_ratio: 1.0,
            z_near: -0.1,
            z_far: -100.0,
            width: 1024,
            height: 1024,
            frame_buffer: Vec::default(),
            pos_buffer: Vec::default(),
            normal_buffer: Vec::default(),
            diffuse_buffer: Vec::default(),
            specular_buffer: Vec::default(),
            glow_buffer: Vec::default(),
            z_buffer: Vec::default(),
            models: Vec::default(),
            lights: Vec::default(),
            shader: crate::rasterization::shader::default_texture_shader,
        }
    }
}
impl Scene {
    pub fn new(width: i32, height: i32) -> Self {
        Scene {
            eye_pos: Vector3f::new(0.0, 0.0, 0.0),
            view_dir: Vector3f::new(0.0, 0.0, -1.0),
            fov: 45.0,
            aspect_ratio: 1.0,
            z_near: -0.1,
            z_far: -100.0,
            width: width,
            height: height,
            frame_buffer: Vec::default(),
            pos_buffer: Vec::default(),
            normal_buffer: Vec::default(),
            diffuse_buffer: Vec::default(),
            specular_buffer: Vec::default(),
            glow_buffer: Vec::default(),
            z_buffer: Vec::default(),
            models: Vec::default(),
            lights: Vec::default(),
            shader: crate::rasterization::shader::default_texture_shader,
        }
    }
    pub fn add_model(&mut self, model: Arc<Mutex<Model>>) {
        self.models.push(model);
    }
    pub fn add_light(&mut self, light: Arc<Mutex<dyn LightTrait>>) {
        self.lights.push(light);
    }
    pub fn delete_model(&mut self, target: &Arc<Mutex<Model>>) {
        self.models.retain(|model| !Arc::ptr_eq(model, target));
    }

    pub fn delete_light(&mut self, target: &Arc<Mutex<dyn LightTrait>>) {
        self.lights.retain(|light| !Arc::ptr_eq(light, target));
    }
    pub fn get_index(&self, x: i32, y: i32) -> i32 {
        self.width * y + x
    }
    pub fn get_penumbra_mask_index(&self, x: i32, y: i32) -> i32 {
        self.width * (y / 4) + (x / 4)
    }
    pub fn set_eye_pos(&mut self, eye_pos: Vector3f) {
        self.eye_pos = eye_pos;
    }
    pub fn get_eye_pos(&self) -> Vector3f {
        self.eye_pos
    }
    pub fn set_view_dir(&mut self, view_dir: Vector3f) {
        self.view_dir = view_dir.normalize();
    }
    pub fn get_view_dir(&self) -> Vector3f {
        self.view_dir
    }
    pub fn set_fov(&mut self, fov: f32) {
        self.fov = fov;
    }
    pub fn get_fov(&self) -> f32 {
        self.fov
    }
    pub fn set_aspect_ratio(&mut self, aspect_ratio: f32) {
        self.aspect_ratio = aspect_ratio;
    }
    pub fn get_aspect_ratio(&self) -> f32 {
        self.aspect_ratio
    }
    pub fn set_z_near(&mut self, z_near: f32) {
        self.z_near = z_near;
    }
    pub fn get_z_near(&self) -> f32 {
        self.z_near
    }
    pub fn set_z_far(&mut self, z_far: f32) {
        self.z_far = z_far;
    }
    pub fn get_z_far(&self) -> f32 {
        self.z_far
    }
    pub fn set_width(&mut self, width: i32) {
        self.width = width;
    }
    pub fn get_width(&self) -> i32 {
        self.width
    }
    pub fn set_height(&mut self, height: i32) {
        self.height = height;
    }
    pub fn get_height(&self) -> i32 {
        self.height
    }
    pub fn set_shader(
        &mut self,
        shader: unsafe fn(*mut Scene, *const Vec<*const dyn LightTrait>, i32, i32, i32, i32),
    ) {
        self.shader = shader;
    }
    pub fn get_shader(
        &self,
    ) -> unsafe fn(*mut Scene, *const Vec<*const dyn LightTrait>, i32, i32, i32, i32) {
        self.shader
    }
    pub fn start_render(&mut self) {
        self.frame_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.pos_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.normal_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.diffuse_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.specular_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.glow_buffer.resize(
            (self.width * self.height) as usize,
            Vector3f::new(0.0, 0.0, 0.0),
        );
        self.z_buffer
            .resize((self.width * self.height) as usize, INFINITY);
        self.frame_buffer.fill(Vector3f::new(0.7, 0.7, 0.7));
        self.z_buffer.fill(INFINITY);
        let modeling = Matrix4f::identity();
        let view = get_view_matrix(self.eye_pos, self.view_dir);
        let projection =
            get_projection_matrix(self.fov, self.aspect_ratio, self.z_near, self.z_far);
        let mvp = projection * view * modeling;
        let mv = view * modeling;

        let scene_ptr = MutScenePtr(self as *const Scene as *mut Scene);

        let mut locked_lights = Vec::with_capacity(self.lights.len());
        let mut raw_const_lights = Vec::with_capacity(self.lights.len());
        let mut raw_mut_lights = Vec::with_capacity(self.lights.len());
        for light in &mut self.lights {
            let guard = light.lock().unwrap();
            let raw_const_ptr = &*guard as *const dyn LightTrait;
            let raw_mut_ptr = raw_const_ptr as *mut dyn LightTrait;
            locked_lights.push(guard);
            raw_const_lights.push(raw_const_ptr);
            raw_mut_lights.push(raw_mut_ptr);
        }
        let vec_const_light_ptr = MutVecConstLightPtr(&mut raw_const_lights);
        let vec_mut_light_ptr = MutVecMutLightPtr(&mut raw_mut_lights);

        let mut locked_models = Vec::with_capacity(self.models.len());
        let mut raw_const_models = Vec::with_capacity(self.models.len());
        let mut raw_mut_models = Vec::with_capacity(self.models.len());
        for model in &mut self.models {
            let guard = model.lock().unwrap();
            let raw_const_ptr = &*guard as *const Model;
            let raw_mut_ptr = raw_const_ptr as *mut Model;
            locked_models.push(guard);
            raw_const_models.push(raw_const_ptr);
            raw_mut_models.push(raw_mut_ptr);
        }
        let vec_const_model_ptr = MutVecConstModelPtr(&mut raw_const_models);
        let vec_mut_model_ptr = MutVecMutModelPtr(&mut raw_mut_models);
        for light in &mut locked_lights {
            light.look_at(unsafe { vec_mut_model_ptr.get() });
        }
        for model in &mut locked_models {
            model.clip(&mvp, &mv);
            model.to_NDC(self.width as u32, self.height as u32);
        }
        let thread_num = std::cmp::min(self.height, SCENE_MAXIMUM_THREAD_NUM);
        let rows_per_thread = (self.height as f32 / thread_num as f32).ceil() as i32;
        let self_width = self.width;
        let self_height = self.height;
        let mut handles = Vec::with_capacity(thread_num as usize);
        for tid in 0..thread_num - 1 {
            let handle = thread::spawn(move || unsafe {
                let models = vec_const_model_ptr.get();
                let scene: *mut Scene = scene_ptr.get_mut();
                for model in &(*models) {
                    let model_ref = (*model).as_ref().unwrap();
                    model_ref.rasterization(
                        scene,
                        tid * rows_per_thread,
                        0,
                        rows_per_thread,
                        self_width,
                    );
                }
            });
            handles.push(handle);
        }
        let handle = thread::spawn(move || unsafe {
            let models = vec_const_model_ptr.get();
            let scene: *mut Scene = scene_ptr.get_mut();
            for model in &(*models) {
                let model_ref = (*model).as_ref().unwrap();
                model_ref.rasterization(
                    scene,
                    (thread_num - 1) * rows_per_thread,
                    0,
                    self_height - (thread_num - 1) * rows_per_thread,
                    self_width,
                );
            }
        });
        handles.push(handle);
        for handle in handles {
            handle.join().unwrap();
        }
        let box_radius = (4.0 * max(self.width, self.height) as f32 / 1024.0).round() as i32;
        for light in &mut locked_lights {
            light.generate_penumbra_mask(unsafe { scene_ptr.get() });
            light.box_blur_penumbra_mask(box_radius);
        }
        // let start = Instant::now();
        // let duration = start.elapsed();
        // println!("花费了 {}ms", duration.as_millis());
        let mut handles = Vec::with_capacity(thread_num as usize);
        for tid in 0..thread_num - 1 {
            let handle = thread::spawn(move || unsafe {
                let lights = vec_const_light_ptr.get();
                let scene: *mut Scene = scene_ptr.get_mut();
                ((*scene).shader)(
                    scene,
                    lights,
                    tid * rows_per_thread,
                    0,
                    rows_per_thread,
                    self_width,
                )
            });
            handles.push(handle);
        }
        let handle = thread::spawn(move || unsafe {
            let lights = vec_const_light_ptr.get();
            let scene: *mut Scene = scene_ptr.get_mut();
            ((*scene).shader)(
                scene,
                lights,
                (thread_num - 1) * rows_per_thread,
                0,
                self_height - (thread_num - 1) * rows_per_thread,
                self_width,
            )
        });
        handles.push(handle);
        for handle in handles {
            handle.join().unwrap();
        }
    }
    pub fn save_to_file(&self, path: &std::path::Path) -> Result<(), Box<dyn std::error::Error>> {
        assert_eq!(
            self.frame_buffer.len(),
            self.width as usize * self.height as usize
        );
        let mut img_buf: Vec<u8> = Vec::with_capacity((self.width * self.height * 3) as usize);
        for y in 0..self.height {
            for x in 0..self.width {
                let idx = (self.width * (self.height - y - 1) + x) as usize;
                for i in 0..3 {
                    img_buf
                        .push((clamp(self.frame_buffer[idx][i], 0.0, 1.0) * 255.0).round() as u8);
                }
            }
        }
        let img = image::RgbImage::from_raw(self.width as u32, self.height as u32, img_buf)
            .ok_or("Failed to create image from buffer")?;
        img.save(path)?;
        Ok(())
    }

    pub fn get_frame_buffer(&self) -> &Vec<Vector3f> {
        &self.frame_buffer
    }
}
